File: Data Structures & Algorithms/implement-prefix-tree/submission-2.java
class PrefixTree {
    // PrefixTreeNode root;

    // public PrefixTree(){
    //     root = new PrefixTreeNode(' ');
    // }
    class PrefixTreeNode{
        private char value;
        private boolean isWord;
        private PrefixTreeNode[] children;
        public PrefixTreeNode(char val){
            this.value = val;
            this.isWord = false;
            this.children = new PrefixTreeNode[26];
        }
    }
    PrefixTreeNode root;

    public PrefixTree(){
        this.root = new PrefixTreeNode(' ');
    }

    public void insert(String word) {
        PrefixTreeNode curr = root;
        for(char x: word.toCharArray()){
            if(curr.children[x - 'a'] == null){
                curr.children[x - 'a'] = new PrefixTreeNode(x);
            }
            curr = curr.children[x - 'a'];
        }
        curr.isWord = true;
    }

    public boolean search(String word) {
        PrefixTreeNode curr = root;
        for(char x: word.toCharArray()){
            if(curr.children[x - 'a'] == null){
                return false;
            }
            curr = curr.children[x - 'a'];
        }
        return curr.isWord;
    }

    public boolean startsWith(String prefix) {
        PrefixTreeNode curr = root;
        for(char x: prefix.toCharArray()){
            if(curr.children[x - 'a'] == null){
                return false;
            }
            curr = curr.children[x - 'a'];
        }
        return true;
    }
}

File: Data Structures & Algorithms/design-word-search-data-structure/submission-2.py
class Trie:
    def __init__(self):
        self.children = {}
        self.isLast = False

class WordDictionary:

    def __init__(self):
        self.root = Trie()

    def addWord(self, word: str) -> None:
        curr = self.root
        for w in word:
            if(w not in curr.children):
                curr.children[w] = Trie()
            curr = curr.children[w]
        curr.isLast = True

    def search(self, word: str) -> bool:
        def dfs(i, newCurr):
            curr = newCurr
            for j in range(i, len(word)):
                if(word[j] == "."):
                    for posStart in curr.children:
                        if(dfs(j + 1, curr.children[posStart])):
                            return True
                    return False
                else:
                    if(word[j] not in curr.children):
                        return False
                    curr = curr.children[word[j]]
            return curr.isLast
        
        return dfs(0, self.root)
